### PR TITLE
small improvements to studio usage reporting

### PR DESCRIPTION
**Pull request by garypen**
A variety of mainly cosmetic improvements to the recently merged studio
usage and spaceport functionality.

closes: #420

### CODE REVIEW

File: apollo-router/src/apollo_telemetry.rs [hunk 1]
```diff
@@ -285,13 +285,19 @@ impl SpanExporter for Exporter {
          * Process the batch
          */
         for span in batch.iter().filter(|span| span.name == "graphql_request") {
-            tracing::debug!(%span.name, ?span.start_time, ?span.end_time);
-            tracing::debug!("span: {:?}", span);
+            // Time may wander and if we ever receive a span where start_time >
+            // endtime, then we should just drop the span and continue processing
+            // the rest of the batch
+            if span.start_time > span.end_time {
```
Comment (Geal) [inline apollo-router/src/apollo_telemetry.rs:291]: I think tracing guarantees that end is always higher than start (but I can't remember where I saw it)

File: apollo-router/src/apollo_telemetry.rs [hunk 2]
```diff
@@ -367,17 +375,21 @@ impl SpanExporter for Exporter {
...
+fn stats_report_key(op: Option<&opentelemetry::Value>, query: &str) -> String {
+    // This represents a logic error in the router, since we should at
+    // least have "-" as the value for the name if no out of band name
+    // was specified with the query.
+    let mut op_name: String = match op {
         Some(v) => v.as_str().into_owned(),
         None => {
-            tracing::warn!("Could not identify operation name: {}", query);
-            return GRAPHQL_UNKNOWN_OPERATION_NAME.to_string();
+            panic!(
```
Comment (Geal) [inline apollo-router/src/apollo_telemetry.rs:389]: should the agent really panic if the router has invalid behaviour? Here that means the agent will stop reporting but the router still handles traffic, right? Or is there an agent restart somewhere?

File: apollo-router/src/apollo_telemetry.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/apollo_telemetry.rs:291]: > I think tracing guarantees that end is always higher than start (but I can't remember where I saw it)

I feel like it should, but SystemTime definitely doesn't and I can't find a reference either. So, better safe than sorry...

File: apollo-router/src/apollo_telemetry.rs [hunk 3]
```diff
@@ -392,25 +404,43 @@ fn normalize(op: Option<&opentelemetry::Value>, query: &str) -> String {
...
-        .into_iter()
-        .filter(|x| {
+    // If we haven't specified an out of band name, then return true
+    // for every operation definition and update the op_name if
+    // we have one.
+    // If we do have an out of band name, then check for equality
+    // with the operation definition name.
+    // If we find more than one match, then in either case we will
+    // fail.
+    let filter: Box<dyn FnMut(&ast::Definition) -> bool> = if op_name == "-" {
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/apollo_telemetry.rs:414]: I originally had a single closure, but it ended up looking neater if I separated the logic into two alternative filters. I wanted to try and keep it as clear as possible. I'll try your suggestion and see if it looks better.

File: apollo-router/src/apollo_telemetry.rs (hunk 2 above)
Comment (garypen) [reply apollo-router/src/apollo_telemetry.rs:389]: It's `panic!` if there is a coding error. i.e. We can only end up here if someone has changed the code in the router and broken the instrumentation. The expected outcome is the router exits abnormally.

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/apollo_telemetry.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/apollo_telemetry.rs]: I've re-worked this a little to make it clear that we expect to be able to calculate a duration from the start and the end and if we can't we are going to ignore the span.

**Pull request merged**